Add filtering ability

MH15 opened the pull request:
# Overview
In the [discord community](https://discord.com/channels/530598289813536771/530598290417385474/888967566058467359), some people noted that there's now a *lot* of projects on the website, and it may be handy to allow filtering of the languages based on tags in the YAML file. This PR implements a basic version of the filtering feature.

Below is a screenshot of the filtering feature in action. I've selected the `compiled` and `jvm` tags, and the list updates accordingly. I've tagged Imp (my language) as `compiled` and `jvm`, and I've tagged C3 as `compiled`. A score is calculated for each language based on the current filtering criteria.

![image](https://user-images.githubusercontent.com/14988198/134390232-634dcfce-b31e-4fe8-b79a-7a68b71a2aff.png)

The usefulness of this feature depends on the quality of tags each contributor can submit for their language. I've tagged a few languages with tags I think seem obvious, but without trying each language I couldn't tag every project.

This feature is implemented in client-side JavaScript, so care was taken to ensure that nothing breaks if JS is turned off. 

# Changes
- `_data/projects.yml` contains a new (optional) field in each project that defines a list of tags.
- `_data/tags.yml` defines the list of tags.
- `index.xhtml` functions identically, but the projects list is moved into its own `div` so it can be sorted. Also, the tag buttons are added.
- `_layouts/default.xhtml` adds a reference to the script.
- `css/style.css` adds styles for the tag buttons.

jamestmartin commented on index.xhtml:
```diff
@@ -15,96 +15,113 @@ canonical: /
 <!-- Exceptions: The Libera, Reddit, IRC, and Discord icons. I don't know where those are from. -->
 <section id="projects">
   <h2>Take a look at our projects</h2>
+  <div id="tags" style="display: none;">
+    {% assign unique_tags = site.data.tags | uniq %}
+    {% for tag_name in unique_tags %}
+    <div class="shell tag" onclick='setFilter(this,"{{tag_name}}")'>{{ tag_name }}</div>
+    {% endfor %}
+    <div class="shell tag" onclick='setFilter(this,null)'>none</div>
+
+  </div>
```
`display: none;` does not actually remove the content, it just hides it. It has no effect on browsers which do not understand CSS, such as Lynx. It would be preferable if this content were added by JS in the first place.

jamestmartin commented on index.xhtml:
```diff
@@ -15,96 +15,113 @@ canonical: /
-          {{ project.organization.name | default: project.organization }}<!--
-        {% if project.organization.website %}--></a><!--{% endif %}
-        {% if project.authors %}-->, including <!--{% endif %}
-        {% if project.author %}-->, including <!--{% endif %}-->
-      {% endif %}
-      {% assign num_authors = project.authors | size %}
-      <!-- {% if num_authors < 2 %}-->
+          {{ project.organization.name | default: project.organization }}
+          <!--
+        {% if project.organization.website %}-->
+        </a>
+        <!--{% endif %}
+        {% if project.authors %}-->, including
+        <!--{% endif %}
+        {% if project.author %}-->, including
+        <!--{% endif %}-->
+        {% endif %}
+        {% assign num_authors = project.authors | size %}
+        <!-- {% if num_authors < 2 %}-->
         {% assign author = project.authors[0] | default: project.author %}
         {% if author.website %}<a href="{{ author.website }}">{% endif %}
           {{ author.name | default: author }}
-        {% if author.website %}</a>{% endif %}
-      {% else %}-->
+          {% if author.website %}</a>{% endif %}
+        {% else %}-->
         {% assign authors_except_last = project.authors | reverse | shift | reverse %}
         {% for author in authors_except_last %}
-          {% if author.website %}<a href="{{ author.website }}">{% endif %}
-            {{ author.name | default: author }}<!--
-          -->{% if author.website %}</a>{% endif %},
+        {% if author.website %}<a href="{{ author.website }}">{% endif %}
+          {{ author.name | default: author }}
+          <!--
+          -->{% if author.website %}
+        </a>{% endif %},
```
These weird comments were like that for a reason. HTML folds multiple spaces to one, but it does not remove spaces between tags entirely. This creates a bunch of extraneous spaces which break the site's formatting (e.g. `many , including Paul Bone`. Please revert or correct this portion of the change.

jamestmartin commented:
Also, non-blocking opinions/feature requests:
* The `none` tag is a bit confusing. I think just clicking on the enabled tags to disable them is sufficient.
* It'd be nice if filtering either hid the unselected languages or highlighted the selected languages, so it'd be easier to tell which is which and how many there are.
* `oop` has no members
* the behavior of selecting multiple tags appears to be 'or', not 'and', which is not the behavior I'd expect, and does not order within the selected tags. perhaps it should only be possible to select one tag at a time (selecting a tag deselects previous tags) until this is fixed?

MH15 commented:
I believe I've addressed the issues you brought up. Also, I've removed the `none` tag, faded out unselected languages, and changed the filtering behavior to "and" instead of "or". Good catch on the Lynx thing, I didn't think of that at all. There's still an empty div in the html (to store the tags as a data attribute) but no text is generated.

jamestmartin commented:
Awesome. I'll let this rest a day before squashing and merging to give other people a chance to request changes.

After the PR is merged I'll add the (many) missing tags to the existing projects list, and fix the remaining issues with extraneous spaces, formatting, and CSS (which only affect the alternate stylesheets).

Thanks for the contribution.

athas commented:
js is inevitable

PaulBone commented:
This looks amazing! :+1:

jamestmartin commented:
This was merged in 0a23f9a89fbb6585918c7bfb46f0d27bd159d423 and was briefly visible on the live site. However, it's been temporarily disabled until we can (correctly) tag more of the languages and I can fix a few bugs with the script. Even though this PR has technically been merged, I'll leave it open until the feature is functional.
Ping me if I forget to take care of this.
